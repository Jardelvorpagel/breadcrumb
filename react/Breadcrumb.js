import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { injectIntl, intlShape } from 'react-intl'
import { Link } from 'render'

const CSS_CLASSES = {
  BREADCRUMB: 'vtex-breadcrumb pa4',
  LINK: 'vtex-breadcrumb__link dib pv1 link light-primary hover-primary ph2',
}

/**
 * Breadcrumb Component.
 */
class Breadcrumb extends Component {
  static propTypes = {
    intl: intlShape.isRequired,
  }

  getPageHint(categories) {
    switch (categories.length) {
      case 1:
        return 'd'
      case 2:
        return 'c'
      case 3:
        return 'sc'
    }
  }

  getCategories(categories) {
    const categoriesSorted = categories
      .slice()
      .sort((a, b) => a.length - b.length)
    return categoriesSorted.map(category => {
      const categoryStripped = category.replace(/^\//, '').replace(/\/$/, '')
      const categories = categoryStripped.split('/')
      const [categoryKey] = categories.reverse()
      const hint = this.getPageHint(categories)
      return {
        key: categoryKey.toLowerCase(),
        value: `${categoryStripped}/${hint}`,
      }
    })
  }

  render() {
    const { slug, productQuery } = this.props

    if (productQuery.loading || !productQuery.product.categories) {
      return null
    }

    const { categories } = productQuery.product

    const categoriesList = this.getCategories(categories)
    return (
      <div className={CSS_CLASSES.BREADCRUMB}>
        <Link className={CSS_CLASSES.LINK} page="store">
          {this.props.intl.formatMessage({ id: 'breadcrumb.home' })}
        </Link>
        {categoriesList.map((category, i) => {
          return (
            <span key={`${category.key}-${i}`}>
              /
              <Link className={CSS_CLASSES.LINK} to={`/${category.value}`}>
                {category.key}
              </Link>
            </span>
          )
        })}
        /
        <span className="ph2"> {slug} </span>
      </div>
    )
  }
}

Breadcrumb.propTypes = {
  /** Search term used **/
  search: PropTypes.string,
  /** Product's slug**/
  slug: PropTypes.string,
  /** Product data**/
  productQuery: {
    /** Product**/
    product: PropTypes.shape({
      /** Categories*/
      categories: PropTypes.arrayOf(PropTypes.string),
    }),
    /** Loading**/
    loading: PropTypes.bool.isRequired,
  },
}

export default injectIntl(Breadcrumb)
