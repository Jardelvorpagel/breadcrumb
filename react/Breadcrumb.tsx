import PropTypes from 'prop-types'
import React, { Component, Fragment, ReactNode } from 'react'
import { Link } from 'vtex.render-runtime'
import unorm from 'unorm'
import { IconHome, IconCaret } from 'vtex.dreamstore-icons'

import breadcrumb from './breadcrumb.css'

const LINK_CLASS_NAME = `${breadcrumb.link} dib pv1 link ph2 c-muted-2 hover-c-link`

interface DefaultProps {
  categories: Array<string>
}
  
interface Props extends DefaultProps {
  term?: string
}

type Category = {
  name: string
  value: string
}

/**
 * Breadcrumb Component.
 */
class Breadcrumb extends Component<Props> {
  public static readonly defaultProps = {
    categories: [],
  }

  public static readonly propTypes = {
    /** Search term or product slug. */
    term: PropTypes.string,
    /** Product's categories. */
    categories: PropTypes.arrayOf(PropTypes.string),
  }

  private get categoriesList(): Array<Category> {
    const { categories } = this.props
    const categoriesSorted = categories
      .slice()
      .sort((a, b) => a.length - b.length)
    return categoriesSorted.map(category => {
      let categoryStripped = category.replace(/^\//, '').replace(/\/$/, '')
      const currentCategories = categoryStripped.split('/')
      const [categoryKey] = currentCategories.reverse()
      categoryStripped = unorm
        .nfd(categoryStripped)
        .toLowerCase()
        .replace(/[\u0300-\u036f]/g, '')
        .trim()
        .replace(/[-\s]+/g, '-')
      categoryStripped += currentCategories.length === 1 ? '/d' : ''
      return {
        name: categoryKey.toLowerCase(),
        value: categoryStripped,
      }
    })
  }

  public render(): ReactNode {
    const { term, categories } = this.props

    if (!categories.length) {
      return null
    }

    return (
      <div className={`${breadcrumb.container} dn db-ns pv3`}>
        <Link className={LINK_CLASS_NAME} page="store.home">
          <IconHome size={16}/>
        </Link>
        {this.categoriesList.map(({ name, value }, i) => (
          <span key={`category-${i}`} className={`${breadcrumb.arrow} ph2 c-muted-2`}>
            <IconCaret orientation="right" size={8}/>
            <Link className={LINK_CLASS_NAME} to={`/${value}`}>
              {name}
            </Link>
          </span>
        ))}

        {term && (
          <Fragment>
            <span className={`${breadcrumb.arrow} ph2 c-muted-2`}>
              <IconCaret orientation="right" size={8}/>
            </span>
            <span className={`${breadcrumb.term} ph2 c-on-base`}>
              {term}
            </span>
          </Fragment>
        )}
      </div>
    )
  }
}

export default Breadcrumb
