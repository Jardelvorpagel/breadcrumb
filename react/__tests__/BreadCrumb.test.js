/* eslint-env jest */
import React from 'react'
import { render } from 'react-testing-library'

import Breadcrumb from '../Breadcrumb'

describe('<BreadCrumb /> component', () => {
  const renderComponent = customProps => {
    const props = {
      term: 'test',
      categories: ['/Eletrônicos/Smartphones/', '/Eletrônicos/'],
      ...customProps,
    }
    return render(<Breadcrumb {...props} />)
  }

  it('should be rendered', () => {
    expect(renderComponent()).toBeDefined()
  })

  it('should match the snapshot', () => {
    expect(renderComponent()).toMatchSnapshot()
  })
})
