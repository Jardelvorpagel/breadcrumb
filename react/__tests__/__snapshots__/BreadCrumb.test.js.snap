// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BreadCrumb /> component should match the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container dn db-ns pv3"
      >
        <a>
          <div
            size="26"
          >
            Icon Home
          </div>
        </a>
        <span
          class="arrow ph2 c-muted-2"
        >
          <div
            orientation="right"
            size="8"
          >
            Icon Caret
          </div>
          <a>
            eletrônicos
          </a>
        </span>
        <span
          class="arrow ph2 c-muted-2"
        >
          <div
            orientation="right"
            size="8"
          >
            Icon Caret
          </div>
          <a>
            smartphones
          </a>
        </span>
        <span
          class="arrow ph2 c-muted-2"
        >
          <div
            orientation="right"
            size="8"
          >
            Icon Caret
          </div>
        </span>
        <span
          class="term ph2 c-on-base"
        >
          test
        </span>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container dn db-ns pv3"
    >
      <a>
        <div
          size="26"
        >
          Icon Home
        </div>
      </a>
      <span
        class="arrow ph2 c-muted-2"
      >
        <div
          orientation="right"
          size="8"
        >
          Icon Caret
        </div>
        <a>
          eletrônicos
        </a>
      </span>
      <span
        class="arrow ph2 c-muted-2"
      >
        <div
          orientation="right"
          size="8"
        >
          Icon Caret
        </div>
        <a>
          smartphones
        </a>
      </span>
      <span
        class="arrow ph2 c-muted-2"
      >
        <div
          orientation="right"
          size="8"
        >
          Icon Caret
        </div>
      </span>
      <span
        class="term ph2 c-on-base"
      >
        test
      </span>
    </div>
  </div>,
  "debug": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllBySelectText": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getAllByValue": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getBySelectText": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "getByValue": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllBySelectText": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryAllByValue": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryBySelectText": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "queryByValue": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
